Add Shipping settings to Cart block

senadir opened the pull request:
Fixes #1488.

- I've added two toggle options, show calculator and hide costs.
- I've also moved `getTotalRowsConfig` inside `FullCart` since it depends on an attribute passed from the above block, we can keep it out and just add an extra param to `getTotalRowsConfig`.
- Hide the options on Empty cart mode.
- Render those changes to the frontend
### Screenshots
![image](https://user-images.githubusercontent.com/6165348/72364267-833eb100-36f6-11ea-8b48-af613402df8d.png)
![image](https://user-images.githubusercontent.com/6165348/72364280-8a65bf00-36f6-11ea-801a-5765522d4cd9.png)
![image](https://user-images.githubusercontent.com/6165348/72364293-92256380-36f6-11ea-9b83-6be63d74ac0f.png)
![image](https://user-images.githubusercontent.com/6165348/72364412-c4cf5c00-36f6-11ea-95a7-675cd0e22309.png)

senadir commented:
One note that I didn’t check yet, if we’re hiding the shipping costs, should we do something about removing it from the API response? or should we leave mismatching values?

Example of mismatched values
![image](https://user-images.githubusercontent.com/6165348/72364667-2c85a700-36f7-11ea-82dd-ce91a0db669f.png)

Aljullu commented:
@senadir is this ready for review? I see it's in the _Review/QA_ pipeline and it has the _Needs review_ tag but it's a draft PR.

senadir commented:
@Aljullu yes and no, while it doesn’t link the option source to wcSettings yet, I would love if you can review the settings boilerplate in general

see
>Fixes #1488 I'm not yet fully sure how we should wrap or present those settings or how to handle global default values, so until we finish the discussion in #1488 I'm keeping this a draft.

Aljullu commented on assets/js/blocks/cart-checkout/cart/edit.js:
```diff
@@ -18,13 +23,59 @@ import EmptyCart from './empty-cart';
+						'woo-gutenberg-products-block'
+					) }
+					help={ __(
+						'Allow customers to estimate shipping',
```
I think `help` texts should end with a period.

haszari replied:
Agreed - the core blocks all use sentence case and full stop for the help text, we should be consistent.

haszari replied:
Also, I find this one a little confusing, are the customers "estimating" shipping or selecting a shipping option (newbie question)?

senadir replied:
estimating shipping should allow users to enter an address and see what the cost for that place is with each shipping method

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/index.js:
```diff
@@ -110,9 +50,77 @@ const cartTotals = {
-const Cart = () => {
-	const currency = getCurrencyFromPriceResponse( cartTotals );
-	const totalRowsConfig = getTotalRowsConfig( cartTotals );
+const Cart = ( { isShippingCalculatorEnabled, isShippingCostHidden } ) => {
```
Is there some whitespace issue that you're fixing here too? This diff is really big. (I reviewed with `Hide whitespace changes` for now.)

senadir replied:
It’s mostly moving the `getTotalRowsConfig` function to inside the Cart component since it depends on attributes based to Cart, mainly `isShippingCostHidden`

haszari replied:
Ah that makes sense :)

haszari replied:
I have a slight preference for adding the param to `getTotalRowsConfig`, so its dependencies are explicit, and less impact on the code. Not a blocker, happy for this to merge as is :)

(There's some fun merging coming with this and #1741 both close to merge!)

haszari commented on assets/js/blocks/cart-checkout/cart/edit.js:
```diff
@@ -18,13 +23,59 @@ import EmptyCart from './empty-cart';
+				/>
+				<ToggleControl
+					label={ __(
+						'Hide shipping costs',
```
Agree with @Aljullu about the two settings being both positive (both "show").

However I think these settings are quite confusing, and getting the naming and help text right is quite challenging. cc @garymurray 

If the settings do map to core settings, then we should look at using similar wording (I haven't compared the wording).

haszari commented on assets/js/blocks/cart-checkout/cart/edit.js:
```diff
@@ -18,13 +23,59 @@ import EmptyCart from './empty-cart';
+						'woo-gutenberg-products-block'
+					) }
+					help={ __(
+						'Hides this shipping costs until an address is entered',
```
Typo `this`, design has: `Hide shipping costs until an address is entered`

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/index.js:
```diff
@@ -27,72 +27,12 @@ import CartLineItemsTable from './cart-line-items-table';
-};
-
 // @todo this are placeholders
 const onActivateCoupon = ( couponCode ) => {
```
This feels like it should be a separate PR, adding coupon support to cart.

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/index.js:
```diff
@@ -27,72 +27,12 @@ import CartLineItemsTable from './cart-line-items-table';
 	console.log( 'coupon activated: ' + couponCode );
 };
-const cartTotals = {
+const fetchedCartTotals = {
```
Is this inline test data needed? We have `previews/cart-items.js` which should contain a valid API response. If you need different data for testing, consider editing the preview data.

haszari replied:
Gotcha. Could potentially add this test data to a new `previews/cart` to match the `store/cart` endpoint. But yep I see that I'm commenting on stuff that came in a previous PR, apologies :)

haszari replied:
FYI @senadir I've shifted the preview cart totals data in

senadir replied:
I can help you rebase if this was merged before your, or I can rebase mine, no problem

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/index.js:
```diff
@@ -110,9 +50,77 @@ const cartTotals = {
+		}
+		const totalDiscount = parseInt( cartTotals.total_discount, 10 );
+		if ( totalDiscount > 0 ) {
+			const totalDiscountTax = parseInt(
```
Should this be in this PR? Looks unrelated to shipping options to me :)

senadir replied:
This was here when I started editing, it only shows up as edited since I moved the function from outside the component to inside it, the changes I made in `assets/js/blocks/cart-checkout/cart/full-cart/index.js` are adding two props, and adding a condition to certain items according to those props.

garymurray commented:
To clarify what each setting does:

**Shipping calculator**

<img width="358" alt="Screenshot 2020-01-16 at 09 36 42" src="https://user-images.githubusercontent.com/10805911/72512302-ba6ea880-3843-11ea-9e05-cd7810bf97a7.png">

This one removes the shipping calculator from the cart completely, which makes the setting below it not needed, so if you disabled this, the setting below (Shipping costs) should be removed.

**Shipping costs**

<img width="347" alt="Screenshot 2020-01-16 at 09 38 03" src="https://user-images.githubusercontent.com/10805911/72512392-e9851a00-3843-11ea-8808-8acdfc13b67f.png">

If this setting is enabled, then the shipping costs are not shown automatically - the customer has to first enter and address to get the shipping costs to display.

senadir commented:
> I wonder if we should take the chance to make both settings positive, so instead of Hide shipping costs it's Shipping costs:
![image](https://user-images.githubusercontent.com/6165348/72519887-8ea9ed80-3858-11ea-900c-125d6cbfe009.png)

Definitely! I remember I brought this up once, double negation seems hard to comprehend cc @garymurray

senadir commented:
I addressed some comments here, and replied to other, this PR still needs some work, like showing shipping costs once an addressed is entred, once Albert merges his shipping calculator PR, we can something like this, btw, the logic is really getting complicated once we start to juggle so many booleans

```js
const [ showShippingCosts, setShowShippingCosts ] = useState( ! isShippingCostHidden );

useEffect( () => {
	if ( isShippingCalculatorEnabled ) {
		if ( isShippingCostHidden ) {
			if ( shippingAddress ) {
				return setShowShippingCosts( true );
			}
		} else {
			return setShowShippingCosts( true );
		}
	}
	return setShowShippingCosts( false );
}, [ isShippingCalculatorEnabled, isShippingCostHidden, shippingAddress ] );
```

garymurray commented:
@senadir @Aljullu 

> I wonder if we should take the chance to make both settings positive, so instead of Hide shipping costs it's Shipping costs:

> Definitely! I remember I brought this up once, double negation seems hard to comprehend cc @garymurray

Does this align with what you saying:

<img width="363" alt="Screenshot 2020-01-16 at 11 46 26" src="https://user-images.githubusercontent.com/10805911/72522572-eb57d900-3855-11ea-8265-7dd07b5790dc.png">

So by default they both toggle on, if you toggle of the first one, the second one is removed automatically as it no longer applies, but you can optionally leave the first one enabled and disable the second one, which would require the customer to enter their address before seeing the shipping cost.

nerrad commented:
> So by default they both toggle on, if you toggle of the first one, the second one is removed automatically as it no longer applies,

If we want parity with the existing settings, this is not consistent with them.  I found the best way to understand these options is to try the existing ones in the WooCommerce settings with the existing (not new) cart.  I tested with a flat shipping cost (vs a geographical zone), so there may be some variations of behaviour depending on those.

garymurray commented:
> If we want parity with the existing settings, this is not consistent with them.

@nerrad yes, the way they are **shown** is counter to what is in Core currently:

<img width="569" alt="Screenshot 2020-01-16 at 13 41 14" src="https://user-images.githubusercontent.com/10805911/72529805-eb5fd500-3865-11ea-904c-fb16afcbb8c0.png">

But the result is the same - the idea shared [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1561#issuecomment-575115959) was to see if it solved the double negation issue.

The original design was as per the core settings - but in block format.

senadir commented:
The new commits hooks up the settings to Woo Core and render them in the frontend, it also rebases to the latest changes.

We have a problem in which `withFeedback` would jump above the settings after the first change to settings, is this something you can look into @Aljullu

Aljullu commented on assets/js/settings/blocks/constants.js:
```diff
@@ -32,6 +32,11 @@ export const DISPLAY_PRICES_INCLUDING_TAXES = getSetting(
+	'isShippingCalculatorEnabled',
+	true
+);
+export const HIDE_SHIPPING_COST = getSetting( 'isShippingCostHidden', false );
```
I think we should name these constants `IS_SHIPPING_CALCULATOR_ENABLED` and `IS_SHIPPING_COST_HIDDEN`, that would keep them consistent with the settings and make it clear they are booleans.

haszari replied:
I agree, let's use the same convention for naming these, so it's clear that they are a value (not an action) and boolean type.

Aljullu commented on assets/js/blocks/cart-checkout/cart/edit.js:
```diff
@@ -37,14 +90,20 @@ const CartEditor = ( { className } ) => {
 				defaultView={ 'full' }
 				render={ ( currentView ) => (
-					<Fragment>
+					<>
```
I don't have any preference, but we are always explicitly writing `Fragment` in tags. Should we keep it consistent? We could even add an [eslint rule](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-fragments.md) for it.

senadir replied:
I have a preference for fewer keystrokes and fewer imports, but If we're always using `Fragment` then I can switch back to it.

Aljullu replied:
Yeah, makes sense... We could open an issue and discuss it there, I just feel that keeping it consistent makes code easier to scan, whichever way we go.

Aljullu replied:
I know I'm contradicting myself :slightly_smiling_face:  but in [another PR](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1736#discussion_r380131662) we were discussing if the short syntax (`<>`) would be better.

I created an issue to discuss/switch: #1761, this way we don't block current open PRs because of that.

senadir commented:
> It's not clear to me if we agreed to make both settings "positive", Gary's designs look like so but that's not implemented yet, right?

As per Darren and Gary, I feel like parity with Woo Core is more important here, removing double negation  could start from Woo Core and get bubbled to us later.

> We also have the issue of the (change address) string. It belongs to the Shipping Calculator but currently it's hidden with the Shipping Costs toggle. As I said in my previous review, I think it's ok dealing with it in a follow-up, but we should either create a follow-up issue for it or not mark this PR as fixing #1488. Otherwise we will probably forget about it.

I don’t think I'm following, the current behavior will hide `Shipping to location (change address)` when shipping calculator gets disabled, shipping costs do not affect that one

![image](https://user-images.githubusercontent.com/6165348/72815152-7ece4700-3c66-11ea-8f25-6df49b7fb939.png)

Aljullu commented:
I think I didn't pull before re-testing it, sorry for the confusion @senadir! :man_facepalming:

> try toggling off or on a settings

Yes, I could reproduce it now. It looks like the order inside `InspectorControls` changes with the re-render? I don't have a big slot-fill knowledge so I'm not sure if there is an easy solution for it. One approach would be to avoid using the HOC and directly insert the `<FeedbackPrompt>` component in `BlockSettings`, but it might be worth exploring if there are other approaches. In any case, we should open an issue to track that. I can do it if you want.

> I don’t think I'm following, the current behavior will hide `Shipping to location (change address)` when shipping calculator gets disabled, shipping costs do not affect that one

Oh ok, nvm. I was confused about what was the calculator and what were the costs.

One thing that still confuses me: I assumed the `Shipping: $1.23` (first line in the screenshot) was not part of the calculator but part of the shipping costs:
![imatge](https://user-images.githubusercontent.com/3616980/72830728-a252bb00-3c81-11ea-9afc-e09178fd35d3.png)

senadir commented:
> One thing that still confuses me: I assumed the Shipping: $1.23 (first line in the screenshot) was not part of the calculator but part of the shipping costs.

Yeah, that’s part of the testing data, my main point of confusion is, should the API do this, or should the setting code be responsible for the total

mikejolley commented:
@senadir Is this awaiting further review, or revisions? There are some conflicts also.

senadir commented:
this has been rebased and retested, it works fine, nothing is blocking this for now.

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/index.js:
```diff
@@ -66,12 +66,12 @@ const Cart = () => {
 	 * Given an API response with cart totals, generates an array of rows to display in the Cart block.
 	 *
-	 * @param {Object} cartTotals - Cart totals data as provided by the API.
+	 * @param {Object} fetchedCartTotals - Cart totals data as provided by the API.
```
This docblock is out of date, can delete `fetchedCartTotals`, if you keep `getTotalRowsConfig` inside the component :)

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/index.js:
```diff
@@ -104,38 +104,39 @@ const Cart = () => {
-	const totalsCurrency = getCurrencyFromPriceResponse( cartTotals );
-	const totalRowsConfig = getTotalRowsConfig( cartTotals );
+	const totalsCurrency = getCurrencyFromPriceResponse( fetchedCartTotals );
+	const totalRowsConfig = getTotalRowsConfig( fetchedCartTotals );
```
The param is no longer needed (assuming `getTotalRowsConfig` stays inside `Cart` component).

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/index.js:
```diff
@@ -196,35 +199,40 @@ const Cart = () => {
+															) }
+												value={
+																option.price
+															}
```
The indentation here doesn't seem quite right. No biggie if it passes the linter (not a blocker), but might be a hint to refactor to reduce nesting/indentation. There's 8 tabs of indentation when `ShippingRatesControl` starts ... pretty deep :)

Assigning the shipping calculator content to a variable might be cleaner, e.g. 

```js
const shippingCalculationsOptions = (
  <fieldset className="wc-block-cart__shipping-options-fieldset">
  // ...
);
```

haszari commented on src/Assets.php:
```diff
@@ -128,6 +128,8 @@ public static function get_wc_block_data( $settings ) {
 				'productCount'                => array_sum( (array) $product_counts ),
 				'attributes'                  => array_values( wc_get_attribute_taxonomies() ),
+				'isShippingCalculatorEnabled' => 'yes' === get_option( 'woocommerce_enable_shipping_calc' ),
+				'isShippingCostHidden'        => 'yes' === get_option( 'woocommerce_shipping_cost_requires_address' ),
```
Might be clearer to wrap this logic in a function – I'm not sure what the convention is for converting settings to booleans for front end. Not a blocker, just feels like boilerplate that we could standardise. e.g. get_boolean_option() => returns true/false.

nerrad replied:
Let's keep an eye on this, but I'm wary of premature abstraction. I usually follow the [rule of three](https://andrewbrookins.com/technology/the-rule-of-three/) before doing abstractions like this.

nerrad replied:
Another option (that'd be trivial to implement here) is using php's `filter_var`:

```php
filter_var( get_option( 'woocommerce_enable_shipping_calc' ), FILTER_VALIDATE_BOOLEAN )
```

And actually just found out from Mike (via slack) that there is a `wc_string_to_bool` function that could also be used (but `filter_var` might be preferable because it will account for other truthy responses).

haszari replied:
`wc_string_to_bool` looks like the right fit to me :)

senadir replied:
do you think it’s a blocker?

haszari replied:
No - `filter_var` is all good from my point of view. (FYI I've requested changes on recent review because of issue in front end and validation error.)

senadir commented:
so I fixed the issue that you found @haszari and did some refactoring and rebased. Ready for another review

senadir commented:
Okay, so the latest round of commits includes:

- fixing the issue of the attributes for good, the problem is that I wasn’t passing the attributes back in the frontend. e8fb026fcd50e5ce4caddc5bf83b3e192baba050

- Showing the shipping costs once an address is submitted (it only checks that you selected a country). 212a304085e64e6990a07688ea16299aafdac147 and 409ec62ce84ff2472535177d63e66ed2b41af3e6

- Disables all shipping related things if shipping is not enabled c5e941973c3d3128e13cfda32cd33281f9bd7ca5

- fixes a side issue in coupons not being enabled by default 4bb114e2722e5a33150231f329f4b6749aa4ef72

mikejolley commented on assets/js/blocks/cart-checkout/cart/full-cart/index.js:
```diff
@@ -96,33 +125,94 @@ const Cart = ( { cartItems = [], cartTotals = {} } ) => {
+			);
+			totalRowsConfig.push( {
+				label: __( 'Shipping:', 'woo-gutenberg-products-block' ),
+				value: DISPLAY_PRICES_INCLUDING_TAXES
```
This value should be hidden if totals are hidden and there is no address yet, but can be a follow up if you want to log it.

mikejolley commented on assets/js/blocks/cart-checkout/cart/edit.js:
```diff
@@ -18,7 +20,55 @@ import EmptyCart from './empty-cart';
+						} )
+					}
+				/>
+				{ isShippingCalculatorEnabled && (
```
This option should be independent of the calculator option, so you can remove the above check.
